codec/decode.go

```diff
@@ -3,86 +3,101 @@ package codec
 import (
 	"encoding/binary"
 	"errors"
+	"reflect"
 )
 
-// DecodeInteger accepts a byte array representing a SCALE encoded integer and performs SCALE decoding of the
-// int, then returns it
-func DecodeInteger(b []byte) (int64, error) {
-	if len(b) == 1 {
-		return int64(b[0] >> 2), nil
-	} else if len(b) == 2 {
+// DecodeInteger accepts a byte array representing a SCALE encoded integer and performs SCALE decoding of the int
+// if the encoding is valid, it then returns (o, bytesDecoded, err) where o is the decoded integer, bytesDecoded is the
+// number of input bytes decoded, and err is nil
+// otherwise, it returns 0, 0, and error
+func DecodeInteger(b []byte) (o int64, bytesDecoded int64, err error) {
+	// check mode of encoding, stored at 2 least significant bits
+	mode := b[0] & 0x03
+	if mode == 0 { // 1 byte mode
+		return int64(b[0] >> 2), 1, nil
+	} else if mode == 1 { // 2 byte mode
 		o := binary.LittleEndian.Uint16(b) >> 2
-		return int64(o), nil
-	} else if len(b) == 4 {
+		return int64(o), 2, nil
+	} else if mode == 2 { // 4 byte mode
 		o := binary.LittleEndian.Uint32(b) >> 2
-		return int64(o), nil
+		return int64(o), 4, nil
 	}
 
-	var o int64
-	var err error
-
+	// >4 byte mode
 	topSixBits := (binary.LittleEndian.Uint16(b) & 0xff) >> 2
 	byteLen := topSixBits + 4
 
-	if byteLen == 4 {
-		o = int64(binary.LittleEndian.Uint32(b[1 : byteLen+1]))
-	} else if byteLen > 4 && byteLen < 8 {
-		upperBytes := make([]byte, 8-byteLen)
-		upperBytes = append(b[5:byteLen+1], upperBytes...)
-		upper := int64(binary.LittleEndian.Uint32(upperBytes)) << 32
-		lower := int64(binary.LittleEndian.Uint32(b[1:5]))
-		o = upper + lower
+	if len(b) < int(byteLen)+1 {
+		err = errors.New("could not decode invalid integer")
 	}
 
-	if o == 0 {
-		err = errors.New("could not decode invalid integer")
+	if err == nil {
+		bytesDecoded = int64(byteLen) + 1
+		if byteLen == 4 {
+			o = int64(binary.LittleEndian.Uint32(b[1:bytesDecoded]))
+		} else if byteLen > 4 && byteLen < 8 {
+			// need to pad upper bytes with zeros so that we can perform binary decoding
+			upperBytes := make([]byte, 8-byteLen)
+			upperBytes = append(b[5:bytesDecoded], upperBytes...)
+			upper := int64(binary.LittleEndian.Uint32(upperBytes)) << 32
+			lower := int64(binary.LittleEndian.Uint32(b[1:5]))
+			o = upper + lower
+		}
+
+		if o == 0 {
+			err = errors.New("could not decode invalid integer")
+		}
 	}
 
-	return o, err
+	return o, bytesDecoded, err
 }
 
 // DecodeByteArray accepts a byte array representing a SCALE encoded byte array and performs SCALE decoding
-// of the byte array, then returns it.  If it is invalid, return nil and error
-func DecodeByteArray(b []byte) ([]byte, error) {
-	var o []byte
-	var err error
+// of the byte array
+// if the encoding is valid, it then returns (o, bytesDecoded, err) where o is the decoded byte array, bytesDecoded is
+// the number of input bytes decoded, and err is nil
+// otherwise, it returns nil, 0, and error
```

Comment:
please simplify comment. Most of this should be implied by function signature, generally godocs don't have explicit return details like this

Reply:
Should also explicity state if bytesDecoded includes the mode and length or not

Reply:
comment simplified